Don't apply `group id` restrictions when BYOGroups is enabled

ana-vinogradova-camunda opened the pull request:
## Description

When BYOGroups is enabled on a 8.8 cluster, users are still limited to adhere to internal id constraints e.g. when adding a Group to a Role or Tenant. However, as the control over the format of the id lies within the IdP, no restrictions should be enforced.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #41096

ana-vinogradova-camunda committed:
refactor: removed unused code

ana-vinogradova-camunda committed:
feat: skip validation when groups are managed externally

In case when the control over the format of the id lies within the IdP,
no id restrictions should be enforced:
- check is group claim is configured
- if so, do not apply usual id validation pattern

ana-vinogradova-camunda committed:
test: add test to validate the change

npepinpe commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/RoleController.java:
```diff
@@ -369,4 +369,17 @@ private CompletableFuture<ResponseEntity<Object>> removeMemberFromRole(
                 .removeMember(request));
   }
+
+  private boolean hasGroupClaimConfigured() {
```
🙃 It's sort of what we want to know, in the end:
```suggestion
  private boolean areGroupsManagedExternally() {
```

npepinpe reviewed (commented):
Direction looks good

ana-vinogradova-camunda committed:
refactor: make check name more clear

ana-vinogradova-camunda committed:
fix: skip validation when groups are managed externally

ana-vinogradova-camunda committed:
test: add test to validate external groups are assigned correctly

ana-vinogradova-camunda commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/RoleController.java:
```diff
@@ -369,4 +369,17 @@ private CompletableFuture<ResponseEntity<Object>> removeMemberFromRole(
+    }
+    final var groupsClaim = oidcConfiguration.getGroupsClaim();
+    return groupsClaim != null && !groupsClaim.isEmpty();
+  }
```
💭 This method is now present in 2 controllers. I think it is OK, but let me know if you have other ideas.

megglos commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/tenant/TenantController.java:
```diff
@@ -170,11 +171,14 @@ public CompletableFuture<ResponseEntity<Object>> assignMappingRuleToTenant(
   @CamundaPutMapping(path = "/{tenantId}/groups/{groupId}")
   public CompletableFuture<ResponseEntity<Object>> assignGroupToTenant(
       @PathVariable final String tenantId, @PathVariable final String groupId) {
+    final boolean areGroupsManagedExternally = areGroupsManagedExternally();
```
❌  the issue also affects the `unassignGroupFromTenant` aka Delete endpoint, that one also needs to get adjusted

ana-vinogradova-camunda committed:
refactor: move the check to `AuthenticationConfiguration`

megglos commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/RoleController.java:
```diff
@@ -263,11 +260,14 @@ public CompletableFuture<ResponseEntity<Object>> assignRoleToClient(
             EntityType.GROUP,
-            securityConfiguration.getCompiledIdValidationPattern())
+            areGroupsManagedExternally
+                ? Pattern.compile(".*", Pattern.DOTALL)
```
❌  that will also disable validation of the roleId, wouldn't it? `validateMemberRequest` uses the pattern for both.

ana-vinogradova-camunda replied:
That's a very good catch, thanks! I'd propose to introduce another validation method for Groups as current validation method is also used for clients, users and mapping rules.

ana-vinogradova-camunda committed:
refactor: extract external ID pattern

megglos reviewed (changes requested):
Happy this gets addressed right away!

Requesting changes for:
* DELETE endpoints need the same fix
* current state also disabled roleId validation in RoleController, TenantController is not affected, as there a dedicated validation pattern is used in `TenantRequestValidator`

ana-vinogradova-camunda commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/RoleController.java:
```diff
@@ -263,11 +260,14 @@ public CompletableFuture<ResponseEntity<Object>> assignRoleToClient(
             groupId,
             EntityType.GROUP,
-            securityConfiguration.getCompiledIdValidationPattern())
+            areGroupsManagedExternally
```
I believe it is clear from the variable name why are we allowing all here (`areGroupsManagedExternally` is quite a descriptive name)

ana-vinogradova-camunda committed:
refactor: move group ID pattern to `SecurityConfiguration`

- introduce `getCompiledGroupIdValidationPattern` to `SecurityConfiguration`

ana-vinogradova-camunda committed:
refactor: use `getCompiledGroupIdValidationPattern` in controllers

ana-vinogradova-camunda committed:
fix: fix unintentional disabling of validation for the `roleId`

- `validateMemberRequest` was using the same pattern for groupId and roleId
This is wrong, as in case of external groups we use different patterns.

The change:
- introduce new `toRoleMemberRequest` that accepts 2 pattern:
  - for role ID
  - for member ID
- change `validateMemberRequest` to accept 2 patterns and pass them correctly.

Original `toRoleMemberRequest` stays the same, but passes the same identifierPattern
to `validateMemberRequest`.

ana-vinogradova-camunda committed:
test: update tests

ana-vinogradova-camunda committed:
fix: address external groups issue for `delete` requests

ana-vinogradova-camunda committed:
test: add tests for external groups for `delete` requests

megglos reviewed (approved):
👍  LGTM!
